### ISSUE TITLE
Implement new ODE Cross Trigger to trigger on objects crossing an ODE Area

### ISSUE THREAD

**Issue by rjhowell44** [prominenceai/deepstream-services-library#741]
New ODE Trigger to trigger on the occurrence of a Tracked Object crossing a Line Area. 

The Trigger will track objects by their unique tracking-id maintaining a history of bounding box coordinates as a bounding box trace. The Trigger and Area (which defines the bbox test point to use) generate a line trace from the bounding box trace and test-point which is used to determine if the object crosses the line Area and in what direction. 

The client can specify the minimum frame count required to trigger and the maximum number of trace points (bbox coordinates) to maintain/use/display. Services are provided to control if/how the line trace is overlaid as display meta for an OSD to display.  

```C
/**
 * @brief Cross trigger that tracks Objects and triggers on the occurrence that an 
 * object crosses the trigger's Line or Polygon Area.
 * @param[in] name unique name for the ODE Trigger
 * @param[in] source unique source name filter for the ODE Trigger, NULL = ANY_SOURCE
 * @param[in] class_id class id filter for this ODE Trigger
 * @param[in] limit limits the number of ODE occurrences, a value of 0 = NO limit
 * @param[in] min_frame_count setting for the minimum number of past 
 * consective frames on both sides of a line (line or polygon area) to trigger an ODE.
 * @param[in] max_trace_points maximum number of past trace points to maintain for
 * each tracked objects. 
 * @param[in] test_method one of the DSL_OBJECT_TRACE_TEST_METHOD_* constants.
 * @return DSL_RESULT_SUCCESS on success, DSL_RESULT_ODE_TRIGGER_RESULT otherwise.
 */
DslReturnType dsl_ode_trigger_cross_new(const wchar_t* name, 
    const wchar_t* source, uint class_id, uint limit, uint min_frame_count, 
    uint max_trace_points, uint test_method);

/**
 * @brief Gets the current test settings for the named Cross trigger.
 * @param[in] name unique name for the ODE Trigger to query
 * @param[out] min_frame_count current setting for the minimum number of past 
 * consective frames on both sides of a line (line or polygon area) to trigger an ODE.
 * @param[out] max_trace_points current setting for the maximum number of past 
 * trace points to maintain for each tracked objects. 
 * @param[out] test_method one of the DSL_OBJECT_TRACE_TEST_METHOD_* constants.
 * @return DSL_RESULT_SUCCESS on success, DSL_RESULT_ODE_TRIGGER_RESULT otherwise.
 */
DslReturnType dsl_ode_trigger_cross_test_settings_get(const wchar_t* name, 
    uint* min_frame_count, uint* max_trace_points, uint* test_method);
    
/**
 * @brief Sets the max trace-points setting for the named Cross trigger.
 * @param[in] name unique name for the ODE Trigger to update
 * @param[in] min_frame_count new setting for the minimum number of past 
 * consective frames on both sides of a line (line or polygon area) to trigger an ODE.
 * @param[in] max_trace_points new setting for the maximum number of past 
 * trace points to maintain for each tracked objects. 
 * @param[in] test_method one of the DSL_OBJECT_TRACE_TEST_METHOD_* constants.
* @return DSL_RESULT_SUCCESS on success, DSL_RESULT_ODE_TRIGGER_RESULT otherwise.
 */
DslReturnType dsl_ode_trigger_cross_test_settings_set(const wchar_t* name, 
    uint min_frame_count, uint max_trace_points, uint test_method);
    
/**
 * @brief Gets the current trace settings for the named Cross trigger.
 * @param[in] name unique name for the ODE Trigger to query
 * @param[out] enabled true if object trace display is enabled, default = disabled 
 * @param[out] color name of the color to use for object trace display, default = no-color.
 * @param[out] line_width width of the object trace if display is enabled.
 * @return DSL_RESULT_SUCCESS on success, DSL_RESULT_ODE_TRIGGER_RESULT otherwise.
 */
DslReturnType dsl_ode_trigger_cross_view_settings_get(const wchar_t* name, 
    boolean* enabled, const wchar_t** color, uint* line_width);
    
/**
 * @brief Sets the trace settings for the named Cross trigger.
 * @param[in] name unique name for the ODE Trigger to update
 * @param[in] enabled set to true to enable object trace display, false otherwise
 * @param[in] color name of the color to use for object trace display.
 * @param[in] line_width width of the object trace if display is enabled.
 * @return DSL_RESULT_SUCCESS on success, DSL_RESULT_ODE_TRIGGER_RESULT otherwise.
 */
DslReturnType dsl_ode_trigger_cross_view_settings_set(const wchar_t* name, 
    boolean enabled, const wchar_t* color, uint line_width);
```

**Comment by rjhowell44** [comment]
example event data
```
Trigger Name    : person-cross
  Unique ODE Id : 1
  NTP Timestamp : 2022-04-19 09:43:10.183340
  Source Data   : ------------------------
    Inference   : Yes
    Source Id   : 0
    Batch Id    : 0
    Pad Index   : 0
    Frame       : 62
    Width       : 1920
    Heigh       : 1080
  Object Data   : ------------------------
    Occurrences : 1
    Obj ClassId : 2
    Infer Id    : 1557520627
    Tracking Id : 7
    Label       : Person
    Confidence  : 0.894973
    Persistence : 1200
    Direction   : 1
    Left        : 234
    Top         : 459
    Width       : 124
    Height      : 322
  Criteria      : ------------------------
    Class Id    : 2
    Frame Count : 1 out of 1
    Min Width   : 0
    Min Height  : 0
    Max Width   : 0
    Max Height  : 0
    Confidence  : 0.4
    Inference   : No
```

**Comment by rjhowell44** [comment]
complete - merged into v0.24.alpha
